[Android Text Input] Remove Samsung restart input workaround

LongCatIsLooong opened the pull request:
Fixes https://github.com/flutter/flutter/issues/73433, where holding the backspace on a Samsung keyboard only deletes 1 character at a time. To delete more characters you'd have to press the backspace key again.

The Samsung "always-restart-input" workaround doesn't seem to be needed anymore (tested on a Galaxy Tab A SMT510, the [duplication bug](https://github.com/flutter/flutter/issues/31512) didn't happen with this patch applied).

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.
- [ ] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

justinmc commented:
I tried this on my S6 and I think the duplication bug is still happening.  Can you try this?  Using the Samsung Korean keyboard and a text field, type:

  * Type: ㄴㅇ
  * Tap once to move the cursor to the start of the field.
  * Type: ㄴ
  * Result: ㅇㄴㄴㅇ
  * Expected: ㄴㄴㅇ


<img width="200" src="https://user-images.githubusercontent.com/389558/107290493-7b496180-6a1b-11eb-85e7-3b44dd1805fb.png" />
<img width="200" src="https://user-images.githubusercontent.com/389558/107290497-7c7a8e80-6a1b-11eb-9bc0-cdc9acc058cb.png" />

LongCatIsLooong commented:
@justinmc thanks for checking! On my SM-T510 it works fine:

https://user-images.githubusercontent.com/31859944/107293732-07aa5300-6a21-11eb-81cf-f0ceee75ac1f.mp4

justinmc commented:
Hmm, do you have any other Samsung devices lying around?  Here is a list of a bunch of devices we were able to reproduce the original bug on: https://github.com/flutter/flutter/issues/31512#issuecomment-588557109

LongCatIsLooong commented:
unfortunately that's the only Samsung device I have. It's running Samsung Keyboard version 3.3.31.83, maybe an update changed how composing regions are handled?

justinmc commented:
v2.0.01-44 for me.

LongCatIsLooong commented:
I switched to 1.12.13+hotfix8 and still couldn't repro. Looks like when I connect the tablet to download the Korean language support the keyboard updated itself and fixed the composing region bug.

hamdikahloun commented:
Remote samsung devices for debugging and testing : https://developer.samsung.com/remotetestlab/rtlDeviceList.action

justinmc commented on shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:
```diff
@@ -497,7 +498,26 @@ private boolean isRestartAlwaysRequired() {
+    }
+
+    // 3.3.31.83 is a known version that's free of the aforementioned bug.
+    // TODO(LongCatIsLooong): Find the minimum version that has the fix.
```
Do you plan to do this in this PR or later?  Definitely my keyboard v2.0.01-44 has the bug, so it's between these two.

LongCatIsLooong replied:
Unfortunately I am still not able to reset the IME version. I've tried `adb uninstall --user 0`. It did uninstall the keyboard but it still wouldn't let me install an earlier version (says downgrade version).

@hamdikahloun thanks for the pointer! It's pretty handy.

justinmc commented on shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java:
```diff
@@ -339,14 +342,22 @@ public void setTextInputEditingState_alwaysSetEditableWhenDifferent() {
+  public void setTextInputEditingState_alwaysRestartsOnAffectedDevices() {
+    // Initialize a TextInputPlugin with a Samsung keypad.
     ShadowBuild.setManufacturer("samsung");
+    // final ShadowPackageManager packageManager = RuntimeEnvironment.getPackageManager();
```
Commented line that maybe you meant to remove.

justinmc commented on shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java:
```diff
@@ -339,14 +342,22 @@ public void setTextInputEditingState_alwaysSetEditableWhenDifferent() {
+    final PackageInfo info = new PackageInfo();
+    info.packageName = "com.sec.android.inputmethod";
+    info.versionCode = 200000000;
+    packageManager.addPackage(info);
```
It's nice that you were able to fake this 👍

justinmc commented on shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java:
```diff
@@ -382,10 +393,63 @@ public void setTextInputEditingState_alwaysRestartsOnAffectedDevices2() {
         testView, new TextInputChannel.TextEditState("", 0, 0, -1, -1));
 
-    // Verify that we've restarted the input.
+    // Verify that we've NOT restarted the input.
```
I think this comment should have stayed as it was?

justinmc reviewed (approved):
LGTM 👍 

Hopefully old keyboards like mine are rare and getting rarer, and the vast majority of users get a fully-functioning keyboard.

chinmaygarde commented:
This looks good to go. Can we land this please?

stuartmorgan commented on shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:
```diff
@@ -497,7 +498,27 @@ private boolean isRestartAlwaysRequired() {
+              .getContext()
+              .getPackageManager()
+              .getPackageInfo("com.sec.android.inputmethod", 0)
+              .getLongVersionCode();
```
This API was added in API level 28 (Pie) but the check above only requires Lollipop to get to this point. This is crashing on old devices, and blocking everything in the flutter/plugins tree. Please fix or revert this ASAP.

justinmc commented:
There is an issue open about a similar problem on a different keyboard: a Japanese keyboard called Simeji https://github.com/flutter/flutter/issues/72980.
